man/man1/rvstack.1
.TH RVSTACK 1NEMO "9 June 2021"

.SH "NAME"
rvstack \- extract Radius-Velocity diagram from a data cube with selected stacking

.SH "SYNOPSIS"
\fBrvstack\fP in=cube out=map [parameter=value]

.SH "DESCRIPTION"
\fBrvstack\fP takes all points within a (narrow) wedge around the major or minor
axis of an assumed galaxy disk data cube, and stacks these data to
construct a two-dimensional \fIRadius-Velocity map\fP,
centered on the galaxy in position and velocity. The intent is to show the rotation
curve along the major axis, or (Z-axis) outflow along the minor axis. If there
is signal at negative velocities, either there are leading spiral arms, or there
is inflow.

By default it will stack the two sides of the galaxy (see also \fBside=\fP),
so the radial coordinate will run from 0 to some Rmax, and 
in velocity it will run from some -Vmax to Vmax.
This is done in such a way as to expect positive rotation around the Z-axis,
or a positive outflow along the Z axis. For this the
position angle (\fBpa=\fP) has been correctly set to the major axis / receding side for rotation,
and minox axis / receding (near) side for outflow.
.PP
Weak signal rotation and/or bulk outflow could be detected this way, but we offer no guarentee!
If you have many like-minded galaxies, stacking these images may beat the noise down even more.
For subsequent stacking \fIccdstack(1NEMO)\fP will need to be used.
.PP
Options exist to scale the Radius and Velocity to different units, such that
objects can be stacked in a physical space, as opposed to observed space. See
\fBrscale=\fP and \fBvscale=\fP.
.PP
.PP
When \fBgscale=t\fP is set, the coordinates are geometrically corrected as well, as follows:
.nf
.ta +2.5i

Rotation		Outflow
--------		---------
R			R/sin(inc)
V/sin(inc)/cos(theta)	V/cos(inc)

.fi
where \fBtheta\fP is the position angle away from the major axis in the plane of the galaxy.
The radius \fBR\fP has been corrected for this as well.
This can leave some gaps in radius and velocity. See CAVEATS below.
.PP
Points along an arc are all accumulated (technically: averaged), and placed on the same radial grid
as the pixel size in the input data cube. We might need some trick next neighbor smoothing
in order for pixels near the center to have a value.

.SH "PARAMETERS"
.so man1/parameters
.TP 20
\fBin=\fP\fIimage_cube\fP
input data cube, in standard \fIimage(5NEMO)\fP format. The third axis must be velocity, or at least
match the units used in the \fBvlsr=\fP keyword below.
[No default]
.TP
\fBout=\fP\fIimage_map\fP
output R-V map. [No default]
.TP
\fBpa=\fP
position angle along which the wedge is taken. This is normally 
the \fBmajor axis receding side\fP (with mode=rotation) or
the \fBminor axis near side\fP (for mode=outflow), in order get a positive velocity signal in the RV map.
When \fBgscale=t\fP is set, will there be a difference how the radius and velocity
are scaled.
.TP
\fBinc=\fP
inclination angle of the disk [45]
.TP
\fBvsys=\fP
systemic velocity [0]   
.TP
\fBcenter=\fP\fIxpos,ypos\fP
rotation center (in pixel units, where 0,0=lower left). If not supplied, the reference pixel will be used.
Default: reference pixel.
.TP
\fBangle=\fP
full angle around the position angle \fBpa=\fP around which points are accumulated
into the R-V plot.  [Default: 10]
.TP
\fBblank=\fP
Value of the blank pixel to be ignored. [Default: 0.0]
.TP
\fBrscale=\fP
Scaling factor for radius. This would be used to stack galaxies based on some physical
(e.g. \fPkpc\fP) scaling, not geometry, see \fBgscale=\fP below.  [Default: 1]
.TP
\fBvscale=\fP
Scaling factor for velocity. This would be  used to stack galaxies based on a mass scaling,
not geometry, see \fBgscale=\fP below. [Default: 1]
.TP
\fBgscale=f|t\fP
A boolean if to scale for (inclination and position angle based) geometry. If selected,
rotation is scaled by 1/sin(i), outflow by 1/cos(i).
Radius is not scaled for rotation, but by 1/cos(i) for outflow.
Small position angle in the disk plane ("theta") based corrections are also applied.
[Default: f]
.TP
\fBmode=r|o\fP
Select output mode: rotation (r) or outflow (o).   [Default: r]
.TP
\fBside=0|-1|1\fP
By default it will stack both sides (0), but this option will allow you to stack one (1) or the other (-1)
side. [Default: 0]
.TP
\fBtab=t|f\fP
(debugging) write a test table.  [Default: f]
.TP
\fBjiggle=t|f\fP
Debugging spreading signal. Don't use.

.SH "CAVEATS"
Code is still under developement, in particular dealing with units is unresolved.
.PP
A wedge has the disadvantage the the noise decreases inversely proportional to the sqrt(radius),
but has a \fImatched-filter\fP approach to a conic outflow. An alternative would be to have
an option for \fBangle=\fP to be replaced with \fBwidth=\fP.
.PP
The gaps in radius and/or velocity produced with \fBgscale=t\fP can be a nuisance. For stacking
many images, for example using \fIccdstack(1NEMO)\fP,
this should disappear, but otherwise some smoothing/interpolation might be needed.
Not implemented.
.PP
Outflow is hard-coded as outflow along the Z axis. What should the off-axis relationship
be? Conical? No theta correction applied here, as we don't have a model like we have for rotation.

.SH "EXAMPLES"
An example using our standard NGC 6503 cube given in km/s velocity units.

.nf

  rvstack ngc6503.ccd rvma.ccd pa=-60 inc=75 vsys=27 center=164,123 rscale=3600 
  rvstack ngc6503.ccd rvmi.ccd pa=30  inc=75 vsys=27 center=164,123 rscale=3600 

  rvstack ngc6503_83.ccd rvma_83.ccd pa=-60 inc=75 vsys=27 center=129,128 rscale=3600 
  rvstack ngc6503_83.ccd rvmi_83.ccd pa=30  inc=75 vsys=27 center=129,128 rscale=3600 

.fi

Here is an example made with \fBmk_cone\fP
.nf
    $NEMO/scripts/csh/mk_cone out=cone2 rmin=0 vscale=0 outflow=0.5 cone=10 flow=rise noise=1
    
.fi

.SH "SEE ALSO"
ccdstack(1NEMO), ccdzshift(1NEMO), ccdmath(1NEMO), image(5NEMO), rotcur(5NEMO)
.PP
https://github.com/jbjolly/LineStacker
.PP
PyStacker https://arxiv.org/abs/2305.10240

.SH "FILES"
.nf
src/image/rotcur/rvstack.c
scripts/csh/mk_cone - example script to create PPV cubr of disk + conic outflow along Z

https://www.cv.nrao.edu/fits/data/samples/cubes/ngc6503.fits  (old 1983 VLA data)
https://...   (TBD - new data)
.fi

.SH "AUTHOR"
Peter Teuben

.SH "UPDATE HISTORY"
.nf
.ta +1.50i +4.0i
6-may-2021	V0.4 First working version	PJT
31-may-2021	V0.7 better geometry correction	PJT
3-jun-2021	V0.8a clarify near/far side choice	PJT
.fi


src/image/rotcur/rvstack.c
/*
 * RVSTACK:  exact Radius-Velocity diagrams for later stacking
 *
 *  See also:     https://github.com/jbjolly/LineStacker      Jolyy JV et al 2020
 *
 *    5-may-2021  PJT  Drafted, first working version
 *
 * @todo  with rescale or gscale, what should be done with a Jy/beam style option
 *
 */

#include <nemo.h>
#include <image.h>

string defv[] = {
  "in=???\n       input cube",
  "out=???\n      output R-V-map",
  "pa=0\n         position angle of disk",
  "inc=45\n       inclination angle of disk",
  "vsys=0\n       systemic velocity to subtract",
  "center=\n      rotation center (mapcenter if left blank, 0,0=lower left)",
  "angle=10\n     (small) angle around major or minor axis",
  "blank=0.0\n    Value of the blank pixel to be ignored",
  "rscale=1\n     Scaling factor for radius (for output)",
  "vscale=1\n     Scaling factor for velocity (for output)",
  "gscale=f\n     Scale radius and velocity by the appropriate geometric sin/cos factor",
  "mode=rot\n     Rotation (r) or Outflow (o) ",
  "side=0\n       Both (0), or positive (1) or negative (-1) side",
  // "geom=f\n       Off axis Geometric correction as well?",
  "tab=f\n        Write a test table?",
  "jiggle=0\n     Jiggle pixels by this amount to fill gaps when gscale set  **TEST**",
  "VERSION=0.8a\n 3-jun-2021 PJT",
  NULL,
};

string usage="extract Radius-Velocity diagram from a data cube with selected stacking";

imageptr velptr = NULL, outptr = NULL, sumptr = NULL;

real pa, pan, inc, vsys, xpos, ypos, zpos;
real undf;


void nemo_main()
{
  stream denstr, velstr, outstr, tabstr;
  real center[2], cospa, sinpa, cosi, sini, sint, cost, sinp, cosp,
       x, y, v, xt, yt, r, vmul, phi, dphi, theta;
  real vr, wt, frang, dx, dy, dz;
  real angle;
  real rscale = getrparam("rscale");
  real vscale = getrparam("vscale");
  int  side   = getiparam("side");
  real dmin, dmax;
  int i, j, k, nx, ny, nz, nundf, nout, nang;
  int ir, iv;
  string mode;
  bool Qrot;
  bool gscale = getbparam("gscale");
  // bool Qgeom = getbparam("geom";)
  bool Qgeom = gscale;
  bool Qtab = getbparam("tab");
  real jiggle = getrparam("jiggle");

  if (jiggle > 0)
    warning("jiddle this is an experiment, it seems not to be useful");

  velstr = stropen(getparam("in"),"r");
  read_image(velstr,&velptr);
  nx = Nx(velptr);
  ny = Ny(velptr);
  nz = Nz(velptr);

  outstr = stropen(getparam("out"),"w");

  mode = getparam("mode");
  Qrot = (*mode == 'r');

  if (hasvalue("center")) {
    if (nemoinpd(getparam("center"),center,2) != 2)
      error("need two values for center=");
    xpos = center[0];
    ypos = center[1];
  } else {
    xpos = Xref(velptr);   // reference pixel
    ypos = Yref(velptr);
  }
  pa = getdparam("pa") * PI/180;
  inc = getdparam("inc") * PI/180;
  angle = getdparam("angle") * PI/180 * 0.5;
  vsys = getdparam("vsys");
  undf = getdparam("blank");

  cospa = cos(pa);
  sinpa = sin(pa);
  sini  = sin(inc);
  cosi  = cos(inc);
    
  nundf = nout = nang = 0;

  dx = Dx(velptr);    
  dy = Dy(velptr);    
  dz = Dz(velptr);

  zpos = (vsys - Zmin(velptr)) / Dz(velptr) + Zref(velptr);

  dprintf(0,"cube:   %d,%d,%d\n",nx,ny,nz);
  dprintf(0,"center: %g,%g,%g\n",xpos,ypos,zpos);
  dprintf(0,"cdelt:  %g,%g,%g\n",dx,dy,dz);

  create_image(&outptr, nx, nz);
  Xmin(outptr) = 0.0; 
  Xref(outptr) = -0.5;
  Dx(outptr)   = ABS(Dx(velptr)) * rscale;
  Namex(outptr)= "RADIUS";
  Namey(outptr)= Namez(velptr);
  Ymin(outptr) = 0.0;
  Yref(outptr) = (Ny(outptr)-1)/2.0;
  Dy(outptr)   = ABS(Dz(velptr)) * vscale;
  Axis(outptr) = 1;
  Object(outptr) =   Object(velptr);
  create_image(&sumptr, nx, nz);
  dprintf(0,"Output map: %d x %d    %g %g %g x %g %g %g\n",
	  Nx(outptr), Ny(outptr),
	  Xmin(outptr), Dx(outptr), Xref(outptr),
	  Ymin(outptr), Dy(outptr), Yref(outptr));
	 

  for (j=0; j<ny; j++) {         // loop over all points in the map
    y = (j-ypos)*dy;
    if (jiggle>0) y += grandom(0,jiggle*dy);
    for (i=0; i<nx; i++) {
      x = (i-xpos)*dx;
      if (jiggle>0) x += grandom(0,jiggle*dy);      

      if (dx > 0)
	phi = atan2(-x,y);       // math convention (ccdgen can create these)
      else
	phi = atan2(x,y);        // astronomical convention of PA
	
      if (Qrot) {                // rotation
	dphi = phi - pa;
	dphi = SYM_ANGLE(dphi);
	if (ABS(dphi) < angle && (side >=0) ) 
	  vmul = 1.0;
	else {
	  dphi += PI;
	  dphi = SYM_ANGLE(dphi);	  
	  if (ABS(dphi) < angle && (side <=0))
	    vmul = -1.0;
	  else
	    continue;
	}
      } else {                   // outflow
	dphi = phi - pa;
	dphi = SYM_ANGLE(dphi);	
	if (ABS(dphi) < angle && (side >=0)) 
	  vmul = -1.0;
	else {
	  dphi += PI;
	  dphi = SYM_ANGLE(dphi);	  
	  if (ABS(dphi) < angle && (side <= 0))
	    vmul = 1.0;
	  else
	    continue;
	}
      }
      // gather some geometric correction factors (theta in the plane, phi on the sky)
      cost = cos(atan(tan(dphi)/cosi));
      sinp = sin(dphi);
      cosp = cos(dphi);
      
      for (k=0; k<nz; k++) {                  // loop over spectral points
	if (CubeValue(velptr,i,j,k) == undf) {
	  nundf++;
	  continue;
	}
	r = sqrt(x*x + y*y) * rscale;
	v = (k-zpos)*dz;
	if (jiggle>0) v+= grandom(0,jiggle*dz);
	v *= vmul;
	v *= vscale;
	if (gscale) {
	  if (Qrot) {
	    v /= sini;
	    if (Qgeom) {
	      v /= cost;
	      r *= sqrt(cosp*cosp+sinp*sinp/(cosi*cosi));
	    }
	  } else {
	    r /= sini;  // note no geom correction
	    v /= cosi;
	  } // Qrot
	} // gscale 
	ir = (int) (r/Dx(outptr));
	iv = (int) (((v-Ymin(outptr))/Dy(outptr)) + Yref(outptr));
	
	if (Qtab && (k == (int)zpos))
	  printf("%d   %g %g  %g    %d %d %g   %d %d\n",nang, r,v,CubeValue(velptr,i,j,k),i,j,dphi,ir,iv);

	if (ir >= Nx(outptr)) continue;
	if (iv >= Ny(outptr)) continue;
	if (ir < 0) continue;	
	if (iv < 0) continue;

	nang++;
	    
	MapValue(outptr,ir,iv) += CubeValue(velptr,i,j,k);
	MapValue(sumptr,ir,iv) += 1.0;
      } // k
    } // i
  } // j

  dmin = MapMax(velptr);
  dmax = MapMin(velptr);
  dprintf(0,"old Cube minmax: %g %g\n", dmax, dmin);
  
  for (j=0; j<Ny(outptr); j++)
    for (i=0; i<Nx(outptr); i++)
      if (MapValue(sumptr,i,j) > 0) {
	MapValue(outptr,i,j) /= MapValue(sumptr,i,j);
	if (MapValue(outptr,i,j) > dmax) dmax = MapValue(outptr,i,j);
	if (MapValue(outptr,i,j) < dmin) dmin = MapValue(outptr,i,j);
      }
  MapMin(outptr) = dmin;
  MapMax(outptr) = dmax;
  dprintf(0,"new Map  minmax: %g %g\n", dmin, dmax);  
  dprintf(0,"%d points selected for %s mode and %s geometry scaled\n",
	  nang, Qrot? "rotation" : "outflow", gscale ? "with" : "not");
  write_image(outstr,outptr);

}

